### PR TITLE
fix: osd stuck with preboot when boot due to OSDMap consumes slowly in advance_pg

### PR DESCRIPTION
**Pull request by jianwei1216**
commit-1
```
- MOSDMap receive 40 epochs osdmap once,
  - store to bluestore
  - then consume_map for every pg
    - OSD::advance_pg call try_get_map
    - try_get_map will acquire map_cache_lock
      - but very high latency
      - 40 epoch osdmaps need cost 1.7~2.5s
- resolve
  - Split map_cache_lock to reduce the lock range
```

commit-2

    fix: osd stuck with preboot when boot due to OSDMap consumes slowly in advance_pg

    question
    - handle_osd_map receives 40 osdmaps and persists them to the osd store
    - after persistence is completed, these 40 osdmaps will be applied to 128 pg
      - 1 osd has 128 pg, 5 shard-queue/thread
      - 1 shard-queue/thread is divided into about 26 pg
      - When pg consumes 40 osdmaps and map_cache misses, it takes about 2 seconds.
      - 26 pg, takes about 52 seconds
        - pg consumes osdmap with too long delay
        - This will cause handle_osd_map to block receiving subsequent osdmaps
        - Makes osd unable to track the latest osdmap epoch
        - This will cause the osd to stay in the preboot state for a long time
        - and even affect business IO.

    Solution: Exchange space for time
    - separate parameters to avoid mutual interference after setting
      - osd_map_bl_cache_size
      - osd_map_bl_inc_cache_size
      - osd_map_cache_size
      - osd_map_cache_burst_size
      - osd_map_cache_reduce_step
    - added osd_map_cache_burst_size configuration parameter
      - When OSD::handle_osd_map receives a new osdmap
        - allows temporarily increasing the max_size of map_cache to cache more osdmaps
      - OSD::tick periodically checks the max_size of map_cache
        - when 128 pg osdmap epochs have caught up with superblock.newest_map
        - restore the max_size of map_cache to osd_map_cache_size
        - avoid taking up too much physical memory

- issue: https://tracker.ceph.com/issues/63609

### CODE REVIEW

**Comment by jianwei1216** [comment]
@ronen-fr  @yuriw @liewegas  @athanatos @neha-ojha  Please REVIEW, thanks

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -1565,28 +1565,31 @@ OSDMapRef OSDService::_add_map(OSDMap *o)
...
+      return retval;
+    }
+    {
+      logger->inc(l_osd_map_cache_miss);
+      epoch_t lb = map_cache.cached_key_lower_bound();
+      if (epoch < lb) {
+        dout(30) << "get_map " << epoch << " - miss, below lower bound" << dendl;
+        logger->inc(l_osd_map_cache_miss_low);
+        logger->inc(l_osd_map_cache_miss_low_avg, lb - epoch);
+      }
```
Comment (Matan-B) [inline src/osd/OSD.cc:1584]: This can be moved into `get_map_cache(epoch)` (See previous comment).

**Review by Matan-B — COMMENTED** [review]
First round of review

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -6266,6 +6266,31 @@ void OSD::tick()
...
+      if (osd_min == 0 || min < osd_min) {
+        osd_min = min;
+      }
+    }
+    if (osd_min == superblock.get_newest_map()) {
+      map_cache_max_size -= cct->_conf->osd_map_cache_reduce_step;
+      if (map_cache_max_size < cct->_conf->osd_map_cache_size) {
+        map_cache_max_size = cct->_conf->osd_map_cache_size;
+      }
+      service.map_cache.set_size(map_cache_max_size);
```
Comment (Matan-B) [inline src/osd/OSD.cc:6283]: I'm not sure that this commit worth the added complexity when there is an already existing configurable.
A user can simply set `osd_map_cache_size` to 500 or to 1024 if the intent is to increase performance over memory usage.

**Comment by rzarzynski** [comment]
CC: @markhpc.

**Comment by jianwei1216** [comment]
> I'm not sure that this commit worth the added complexity when there is an already existing configurable.
> A user can simply set osd_map_cache_size to 500 or to 1024 if the intent is to increase performance over memory usage.

@Matan-B 
The main consideration here is the osd physical memory occupation.
If you adjust the existing osd_map_cache_size parameter to 500,
It will have the following impacts:

1. osd_map_cache_size This parameter will directly affect the 3 osd to lru cache queues.
The item added is not just 500, but 500*3=1500,
```
   map_cache(cct, cct->_conf->osd_map_cache_size),
   map_bl_cache(cct->_conf->osd_map_cache_size),
   map_bl_inc_cache(cct->_conf->osd_map_cache_size),
```
This will increase physical memory consumption,
This is the purpose of splitting osd_map_cache_size into 3 parameters osd_map_cache_size/osd_map_bl_cache_size/osd_map_bl_inc_cache_size.

2. The cache contents of map_cache and map_bl_cache/map_bl_inc_cache are different:
map_cache is for caching decoded OSDMaps,
map_bl_cache/map_bl_inc_cache is for caching OSDMap bufferlist,

The purpose of map_cache is to allow osd to hit OSD::handle_osd_map and OSD::advance_pg in the cache without having to read the OSDMap bufferlist on the HDD.

The purpose of map_bl_cache/map_bl_inc_cache is to share the OSDMap bufferlist between osd-osd and osd-client (OSDService::build_incremental_map_msg)

Therefore, there is no need to force their caches to be bundled together.

3. The purpose of increasing osd_map_cache_burst_size:
(1) When OSDMap changes,
(2) osd received 40 osdmap bufferlists,
(3) OSD::handle_osd_map decoded 40 OSDMap, add to map_cache lru
(4) After storing 40 osdmaps on the HDD disk, they will be delivered to 128 pg events
(5) Each pg must go through OSD::advance_pg to apply 40 osdmaps to the PG state machine, 128 * 40 = 5120
(6) If the old data in map_cache that has not been applied to advance_pg is eliminated by the subsequent new 40 osdmaps,
Each hdd osd has 5 shard-threads, and each thread is allocated 128/5=26 pg.
In a shard-thread, 26 pgs need to schedule OSD::advance_pg to apply 40 osdmaps serially.
You need to constantly read the OSDMap bufferlist on the HDD disk.
The efficiency is very low. When a large number of osdmaps need to be applied, the time length will be increased, showing that the osd is stuck in the preboot state.

Increase osd_map_cache_burst_size = 500 in order to improve the map_cache cache hit rate.
OSD::hande_osd_map will increase the upper limit of map_cache to 500 when receiving map.
Until OSD::tick determines that all 128 pg have been used as the latest newest_map, map_cache will be reduced to osd_map_cache_size=50 according to osd_map_cache_reduce_step=10
In this way, 128 pg can quickly catch up with the latest osdmap

**Pull request closed without merging**